controller/src/main/java/io/pravega/controller/store/stream/records/EpochRecord.java

```diff
@@ -46,21 +45,37 @@
     private final int referenceEpoch;
     private final List<StreamSegmentRecord> segments;
     private final long creationTime;
-    @Getter(AccessLevel.PRIVATE)
     private final Map<Long, StreamSegmentRecord> segmentMap = new HashMap<>();
     
     @Builder
-    EpochRecord(int epoch, int referenceEpoch, List<StreamSegmentRecord> segments, long creationTime) {
+    /**
+     * This is a private constructor that is only directly used by the builder during the deserialization. 
+     * The deserialization passes @param copyCollections as false so that we do not make an immutable copy of the collection
+     * for the collection passed to the constructor via deserialization. 
+     *
+     * The all other constructors, the value of copyCollections flag is true and we make an immutable collection copy of 
+     * the supplied collection. 
+     * All getters of this class that return a collection always wrap them under Collections.unmodifiableCollection so that
+     * no one can change the data object from outside.  
+     */
+    private EpochRecord(int epoch, int referenceEpoch, List<StreamSegmentRecord> segments, long creationTime, boolean copyCollections) {
         this.epoch = epoch;
         this.referenceEpoch = referenceEpoch;
-        this.segments = ImmutableList.copyOf(segments);
+        this.segments = copyCollections ? ImmutableList.copyOf(segments) : segments;
         this.creationTime = creationTime;
         this.segmentMap.putAll(segments.stream().collect(Collectors.toMap(StreamSegmentRecord::segmentId, x -> x)));
```

Comment:
Why is this doing a collect into a map, only to make a copy into another map ?

Reply:
we compute and keep this map from a list so that this data object can efficiently respond to queries for `getSegmentForId`